File: Stack/Monotonic Stack/Remove K Digits/increasing_stack.cpp
#include<bits/stdc++.h>
using namespace std;

//Time complextiy :- O(n)
// Space complexity :- O(n)

// Approach :-  
// using stack maintain a monotonice increasing stack 
// pop the element to make it monotonic until k is valid 
// eliminate all the leading zero and return '0' if nothing exists 


//Link :- https://www.geeksforgeeks.org/problems/remove-k-digits/1

class Solution {
  public:
    string removeKdig(string &s, int k) {
        stack<char> st;
        
        // traverse from left to right on digits 
        for(auto ch: s){
            // maintain a monotonic stack of increasing 
            // until k expires
            while(!st.empty() && k>0 && (st.top()-'0') > (ch - '0')){
                st.pop();
                k--;
            }
            // add current element to the stack
            st.push(ch);
        }
        // remove exactly k element from the stack if not already removed 
        while(k--){
            st.pop();
        }
        string res = "";
        
        // build result from the number left in stack
        while(!st.empty()){
            res += st.top();
            st.pop();
        }
        // reverse the string to get the actual answer 
        reverse(res.begin(), res.end());
        
        // delete all the leading zeros
        while(res.size()>0 && res[0] == '0')
            res = res.substr(1);
        
        return res.size() == 0 ? "0" : res;
    }
};

File: Stack/Monotonic Stack/Subarrays with First Element Minimum/increasing_stack.java
import java.util.*

//Time complexity :- O(n)
//Space complexity :- O(n)

// Approach :- 
// we'll use stack and maintain a monotonic stack in this case 
// with the help of monotonic stack we can track the minimum element 
// and the size of stack will tell us all the element in the stack will contribute to how many arrays 


//Link :- https://www.geeksforgeeks.org/problems/subarrays-with-first-element-minimum/1

class Solution {
    public int countSubarrays(int[] arr) {
        int res = 0;
        Stack<Integer> st = new Stack<>();
        
        for(int num: arr){
            // maintain a monotonic stack
            while(!st.isEmpty() && st.peek() > num){
                st.pop();
            }
            // add current element into the stack
            st.push(num);
            // add length of stack into answer 
            res += st.size();
        }
        return res;
    }
}
